sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldHaveCorrectSignature.cs

```diff
@@ -44,79 +43,111 @@ public sealed class TestMethodShouldHaveCorrectSignature : SonarDiagnosticAnalyz
+                m => GetFaultMessage(m, publicOnly: true, allowGenerics: true)
+            },
+            {
+                KnownType.NUnit_Framework_TestCaseSourceAttribute,
+                m => GetFaultMessage(m, publicOnly: true, allowGenerics: true)
+            },
+            {
+                KnownType.NUnit_Framework_TheoryAttribute,
+                m => GetFaultMessage(m, publicOnly: true, allowGenerics: false)
+            },
+
+            // XUnit
+            {
+                KnownType.Xunit_FactAttribute,
+                m => GetFaultMessage(m, publicOnly: false, allowGenerics: false)
+            },
+            {
+                KnownType.Xunit_TheoryAttribute,
+                m => GetFaultMessage(m, publicOnly: false, allowGenerics: true)
+            },
+            {
+                KnownType.LegacyXunit_TheoryAttribute,
+                m => GetFaultMessage(m, publicOnly: false, allowGenerics: true)
+            }
+        };
+
+        protected override void Initialize(SonarAnalysisContext context) =>
+            context.RegisterSyntaxNodeActionInNonGenerated(AnalyzeMethod, SyntaxKind.MethodDeclaration);
+
+        private void AnalyzeMethod(SyntaxNodeAnalysisContext c)
         {
-            context.RegisterSyntaxNodeActionInNonGenerated(
-                c =>
-                {
-                    var classDeclaration = (ClassDeclarationSyntax)c.Node;
-                    var classSymbol = c.SemanticModel.GetDeclaredSymbol(classDeclaration);
-                    if (classSymbol == null)
-                    {
-                        return;
-                    }
-
-                    var allFaultyMethods = classSymbol.GetMembers()
-                        .OfType<IMethodSymbol>()
-                        .Select(m => new { method = m, testType = ToKnownTestType(m) })
-                        .Where(tuple => tuple.testType != null)
-                        .Select(
-                            tuple =>
-                            new
-                            {
-                                Location = tuple.method.Locations.First(),
-                                Message = GetFaults(tuple.method, tuple.testType).ToSentence()
-                            })
-                        .Where(tuple => tuple.Message != null);
-
-                    foreach (var faultyMethod in allFaultyMethods)
-                    {
-                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, faultyMethod.Location,
-                            faultyMethod.Message));
-                    }
-                },
-                SyntaxKind.ClassDeclaration);
+            var methodSymbol = c.SemanticModel.GetDeclaredSymbol(c.Node) as IMethodSymbol;
+            if (methodSymbol == null)
+            {
+                return;
+            }
+
+            var validator = GetValidator(methodSymbol);
+            var message = validator(methodSymbol);
```

Comment:
This seems like the 2 previous statements could be merged. Ofc you will need to update the signature of `GetValidator`.